### PR TITLE
Handle case of temp:group schema

### PR DESCRIPTION
**Pull request by obriard**
Case of temp:group schema

## Description

Return a schema when doing a GetSchema on a temp:group schema

## Related Issue

Ability to see the content of a list of type Group in the ACC web ui

## Motivation and Context

Schema of type temp content was returned empty

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by junliadobe — APPROVED** [review]

File: src/domUtil.js [hunk 1]
```diff
@@ -342,7 +342,7 @@ class DomUtil {
                     xmlRoot.appendChild(xmlElement);
                 }
                 else if (t == "object") {
-                    if (value.length !== undefined && value.length !== null) {
+                    if (value.length !== undefined && value.length !== null && value.length > 0) {
```
Comment (mkiki) [inline src/domUtil.js:345]: What is the point of this change?

File: src/client.js [hunk 2]
```diff
@@ -1654,16 +1654,41 @@ class Client {
...
         if (!entity) {
-          entity = await this.getEntityIfMoreRecent("xtk:schema", schemaId, "xml", internal);
-          if (entity) {
-            const impls = DomUtil.getAttributeAsString(entity, "implements");
-            if (impls === "xtk:persist" && schemaId !== "xtk:session" && schemaId !== "xtk:persist") {
-                // Ensure xtk:persist is present by loading the xtk:session schema
-                await this.getSchema("xtk:session", "xml", true);
+              // special case of temp schema for nms:group
+              if (schemaId.startsWith("temp:group:")) {
+                  var parts = schemaId.split(":");
```
Comment (mkiki) [inline src/client.js:1659]: Please use const whenever possible

File: src/client.js [hunk 3]
```diff
@@ -1654,16 +1654,41 @@ class Client {
...
+                    "operation": "get",
+                    "select": {
+                      "node": [
+                        { "expr": "@id" },
+                        { "expr": "extension" }
+                      ]
+                    },
+                    "where": {
+                      "condition": [
+                        { "expr": "@id=" + parts[2] }
```
Comment (mkiki) [inline src/client.js:1671]: This is a potential xtk injection my friend. You may want to check that parts[2] is a valid number. You can use XtkCaster.asLong for example. Even if the injection is theoretical, it's important to ALWAYS escape or secure the input parameters of public functions

File: src/client.js [hunk 4]
```diff
@@ -1654,16 +1654,41 @@ class Client {
...
+                      ]
+                    }
+                  };
+                  // Convert to current representation
+                  queryDef = this._convertToRepresentation(queryDef, "SimpleJson");
+                  const query = this.NLWS.xtkQueryDef.create(queryDef);
+                  const groupSchema = await query.executeQuery();
```
Comment (mkiki) [inline src/client.js:1678]: The corresponding SOAP call should have "internal" flag set as this query is performed under the hood by the SDK

File: src/client.js [hunk 5]
```diff
@@ -1654,16 +1654,41 @@ class Client {
...
+                      const impls = DomUtil.getAttributeAsString(entity, "implements");
+                      if (impls === "xtk:persist" && schemaId !== "xtk:session" && schemaId !== "xtk:persist") {
+                        // Ensure xtk:persist is present by loading the xtk:session schema
+                        await this.getSchema("xtk:session", "xml", true);
+                      }
+                      await this._entityCache.put("xtk:schema", schemaId, entity);
+                      await this._methodCache.put(entity);
+                  }
             }
-            await this._entityCache.put("xtk:schema", schemaId, entity);
```
Comment (mkiki) [inline src/client.js:1664]: Is there any reason why the temp:group schemas are not cached?
If so, I think it's important to have a comment in the code that explains why

File: src/client.js [hunk 6]
```diff
@@ -1654,16 +1654,41 @@ class Client {
...
         var entity = await this._entityCache.get("xtk:schema", schemaId);
         if (!entity) {
-          entity = await this.getEntityIfMoreRecent("xtk:schema", schemaId, "xml", internal);
-          if (entity) {
-            const impls = DomUtil.getAttributeAsString(entity, "implements");
-            if (impls === "xtk:persist" && schemaId !== "xtk:session" && schemaId !== "xtk:persist") {
-                // Ensure xtk:persist is present by loading the xtk:session schema
-                await this.getSchema("xtk:session", "xml", true);
+              // special case of temp schema for nms:group
+              if (schemaId.startsWith("temp:group:")) {
```
Comment (mkiki) [inline src/client.js:1658]: We'll need documentation for this new type of schemas

File: test/client.test.js [hunk 7]
```diff
@@ -434,6 +434,42 @@ describe('ACC Client', function () {
...
+                    <desc><![CDATA[]]></desc>
+                    <folder _cs="Lists" fullName="/Profiles and Targets/Lists/" id="1199"/>
+                  </group>
+                  </pdomOutput></ExecuteQueryResponse>
+              </SOAP-ENV:Body>
+              </SOAP-ENV:Envelope>`));
+          
+            var schema = await client.getSchema("temp:group:2200");
+            expect(schema["namespace"]).toBe("temp");
+            expect(schema["name"]).toBe("query");
```
Comment (mkiki) [inline test/client.test.js:470]: Can you add a test (in application.test.js) which uses the getSchema function of the application object and also goes a little bit deeper in the assertions, i.e. which checks that the <key>, <attribute>, etc. are properly returned

File: test/client.test.js [hunk 8]
```diff
@@ -3767,7 +3803,7 @@ describe('ACC Client', function () {
             const schema = client.newSchema(xml);
             const jobs = schema.root.children["jobs"];
             expect(jobs.target).toBe("xtk:job");
-            await expect(jobs.linkTarget()).rejects.toThrow("Cannot read property 'getSchema' of null");
+            await expect(jobs.linkTarget()).rejects.toThrow("Cannot read properties of null (reading 'getSchema')");
```
Comment (mkiki) [inline test/client.test.js:3806]: This is not a good error message. Why is there such a change in the error message: it does not seem related to your change at all and this is suspicious

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi @obriard, there are a couple of issues with this PR, please see the various comments.

File: src/domUtil.js (hunk 1 above)
Comment (obriard) [reply src/domUtil.js:345]: to handle this kind of attribute:
```xml
<element label="label" name="name" >
      <attribute label="Primary key" name="id" />
</element>
```

File: src/client.js (hunk 5 above)
Comment (obriard) [reply src/client.js:1664]: because it fails when calling `put` method of `XtkEntityCache`
```js
const namespace = entity.getAttribute("namespace");
```

Comment (mkiki) [reply src/client.js:1664]: I don't understand. I do not think that's a good reason. Having a different behaviour depending on the schema can be very comfusing and source of bugs, incorrect assumptions for the user. So if we do not cache such schemas, there should be a good reason. In other words we decide and justify : do we WANT to cache such schemas or not. If not, we explain and document. If yes, we make it work in all scenarios, and this means understnading which the put method fails in this case, and find a solution.

File: src/client.js (hunk 6 above)
Comment (mkiki) [reply src/client.js:1658]: Yes, I mean a kind of schema which was not previously handled by the SDK. So it should still be documented. As we also take the opportunity of this SDK to document obscure thinks in Campaign. In this case, it would be great to have a small paragraph explaining what schemas starting with temp:group: are, how they are constructed, and give examples.

File: src/domUtil.js (hunk 1 above)
Comment (mkiki) [reply src/domUtil.js:345]: This function converts JSON to XML. I don't understand to what JSON your example comments or why this needs to be changed. Changes in the JSON <=> XML conversion are potentially very impactful. This would also warrant appropriate unit tests. 
Are we changing the XML created from the JSON is any way? If yes, then. this is a breaking change

File: test/client.test.js (hunk 8 above)
Comment (mkiki) [reply test/client.test.js:3806]: Sorry, that' still not good. First, it seems to work perfectly on 1.1.24, so this change is probably not necessary
If not can you explain why you changed the test vs. the code. AFAIK, it a test fails, it's s sign of a potential regression. Changing the test merely hides the issue and does not necessarily fixes it. Could you explain the reason of this failure and why you decided to change the test instead of the code?

File: test/domUtil.test.js [hunk 9]
```diff
@@ -153,7 +153,7 @@ describe('DomUtil', function() {
             assert.strictEqual(fromJSON({ a: { "$": "Hello" } }), '<root><a>Hello</a></root>');
             assert.strictEqual(fromJSON({ a: "World", "$a": "Hello" }), '<root a="World"><a>Hello</a></root>');
             assert.strictEqual(fromJSON({ "a": [ { "i":1 }, { "i": 2 } ] }), '<root><a i="1"/><a i="2"/></root>');
-            assert.strictEqual(fromJSON({ "a": [ ] }), '<root/>');
```
Comment (mkiki) [inline test/domUtil.test.js:156]: That is a breaking change.

File: test/client.test.js (hunk 8 above)
Comment (mkiki) [reply test/client.test.js:3806]: In any cases, it should be a separate PR as it is not related to the temp:group schema support

**Review by mkiki — CHANGES_REQUESTED** [review]

File: test/client.test.js (hunk 8 above)
Comment (obriard) [reply test/client.test.js:3806]: Currently we observe this regression on acc web ui :
image

Comment (obriard) [reply test/client.test.js]: I have reverted the change on the test. It is a regression that has present at least since version 1.1.23 : I tested on versions 1.1.24 and  1.1.25 the test fails.

File: src/client.js (hunk 5 above)
Comment (obriard) [reply src/client.js:1664]: I fix the put method failure, temp schemas are now cached

File: src/client.js [hunk 10]
```diff
@@ -1740,16 +1740,51 @@ class Client {
...
-            const impls = DomUtil.getAttributeAsString(entity, "implements");
-            if (impls === "xtk:persist" && schemaId !== "xtk:session" && schemaId !== "xtk:persist") {
-                // Ensure xtk:persist is present by loading the xtk:session schema
-                await this.getSchema("xtk:session", "xml", true);
+              // special case of "temp:group:*" schemas for nms:group
+              // Schema "temp:group:*" is not cached because life cycle of this kind of schema is not the same as the others schemas
+              if (schemaId.startsWith("temp:group:")) {
+                  const parts = schemaId.split(":");
+                  var queryDef = {
+                    "schema": "nms:group",
```
Comment (mkiki) [inline src/client.js:1748]: Use const

File: src/client.js [hunk 11]
```diff
@@ -1740,16 +1740,51 @@ class Client {
...
+                  // Convert to current representation
+                  queryDef = this._convertToRepresentation(queryDef, "SimpleJson", "xml");
+                  const query = this.NLWS.xml.xtkQueryDef.create(queryDef);
+                  try {
+                      const groupSchema = await query.executeQuery(internal);
+                      const extension = DomUtil.findElement(groupSchema, "extension");
+                      if (extension) {
+                          entity = extension;
+                      } else {
+                          entity = null;
```
Comment (mkiki) [inline src/client.js:1771]: You should test the error code and return null if the group or schema is not found only. There are all kind of exceptions, such as network issues, token expiration, etc. and we want those to be reported. In case the schema is not found, we return null because that's the contract of the function

**Review by mkiki — CHANGES_REQUESTED** [review]

File: src/client.js [hunk 12]
```diff
@@ -1740,16 +1740,42 @@ class Client {
...
+                    "where": {
+                      "condition": [
+                        { "expr": "@id=" + XtkCaster.asLong(parts[2]) }
+                      ]
+                    }
+                  };
+                  // Convert to current representation
+                  queryDef = this._convertToRepresentation(queryDef, "SimpleJson");
+                  const query = this.NLWS.xtkQueryDef.create(queryDef);
+                  const groupSchema = await query.executeQuery(internal);
```
[the remark this answers is not in the document]
Comment (obriard) [reply src/client.js]: internal is ignore, executeQuery calls _callMethod wich calls _prepareSoapCall  with internal parameter always equals to false
https://github.com/adobe/acc-js-sdk/blob/master/src/client.js#L1847
To marks this call internal , some changes need to be done on _callMethod , I'm not sure we want to do this in this PR.

File: src/client.js (hunk 10 above)
Comment (obriard) [reply src/client.js:1748]: I will use let because queryDef is assign below :
`queryDef = this._convertToRepresentation(queryDef, "SimpleJson", "xml");`

File: src/client.js (hunk 11 above)
Comment (obriard) [reply src/client.js:1771]: I have set entity to null only for CampaignExceptions ,for others exceptions I re-throw the exception.

Comment (mkiki) [reply src/client.js:1771]: No, you have to set entity to null if the group does not exist of if does not have an extension node. CampaignException can represent tons of errors. You have to check actual error code return carefully.

Also, do not forget to have the corresponding unit tests: execption cases needs to be tested too

**Review by mkiki — CHANGES_REQUESTED** [review]

**Review by mkiki — APPROVED** [review]
